fsm.go

```diff
@@ -31,6 +31,24 @@ type FSM interface {
 	Restore(io.ReadCloser) error
 }
 
+// BatchingFSM extends the FSM interface to add an ApplyBatch function. This can
+// optionally be implemented by clients to enable multiple logs to be applied to
+// the FSM in batches. Up to MaxAppendEntries could be sent in a batch.
+type BatchingFSM interface {
+	// ApplyBatch is invoked once a batch of log entries has been committed and
+	// are ready to be applied to the FSM. ApplyBatch will take in an array of
+	// log entries. These log entries could be of a few types clients should
+	// check the log type prior to attempting to decode the data attached.
+	// Presently the LogCommand and LogConfiguration types will be sent.
```

Comment:
It's perhaps obvious but is it worth stating that the batch will always represent the logs in the order they have been committed and will not include any gaps. FSMs implementing this _must_ apply the commands in the batch in the same order (or equivalent to them being applied in same order) to maintain the FSM invariants.